bw6761 golang wrapper

ImmanuelSegol opened the pull request:
## Describe the changes

This PR...

## Linked Issues

Resolves #

ImmanuelSegol committed:
fix cuda test - remove boost (#197)

* refactor: remove boost

ImmanuelSegol committed:
Fix CI - dont use deprecated package name  (#216)

Update icicle crate name in examples and benches

LeonHibnik committed:
add mont_r and mont_r_inv and fix args (#187)

Resolves #186

ImmanuelSegol committed:
update docs (#222)

* refactor

* lint

* Typo and spacing

---------

Co-authored-by: Jeremy Felder <jeremy.felder1@gmail.com>

jeremyfelder committed:
[CI]: Adds C++/CUDA CI and conditional workflow runs (#223)

* Add cmake tests for cpp primitives

* Add cpp/cuda formatting

* Add conditional steps based on files changed for faster required checks

* Update runs-on for check files changed

ImmanuelSegol committed:
setup

ImmanuelSegol committed:
refactor: fix compile issues

DmytroTym commented on icicle/curves/bw6_761/curve_config.cuh:
```diff
@@ -3,6 +3,9 @@
 #include "../../primitives/field.cuh"
 #include "../../primitives/projective.cuh"
 #if defined(G2_DEFINED)
+#include "../../primitives/extension_field.cuh"
```
I don't think you need lines 6-8. I assume they were added to solve a compilation error but the problem was in the other place (in projective.cu). G2 curve isn't over the extension field and we should be able to speed up compilation quite a bit by removing this unnecessary import.

ImmanuelSegol replied:
Done

DmytroTym commented on icicle/curves/bw6_761/curve_config.cuh:
```diff
@@ -21,6 +24,7 @@ namespace BW6_761 {
   typedef Projective<point_field_t, scalar_t, b, gen_x, gen_y> projective_t;
   typedef Affine<point_field_t> affine_t;
 #if defined(G2_DEFINED)
+  typedef ExtensionField<PARAMS_BW6_761::fq_config> g2_point_field_t;
```
This should be removed as well

ImmanuelSegol replied:
Done, i removed the wrong imports as well from 377

DmytroTym commented on icicle/utils/sharedmem.cuh:
```diff
@@ -62,6 +62,12 @@
 #include "../curves/bls12_381/curve_config.cuh"
 #include "../curves/bn254/curve_config.cuh"
 
+__device__ void Error_UnsupportedType() 
```
@vhnatyk pls take a look, is this how `Unresolved extern function '_Z21Error_UnsupportedTypev'` should be resolved? Maybe `#include "../curves/bw6_671/curve_config.cuh"` can help?

vhnatyk replied:
while seems this may trick a linker - for compatibility at least with ntt templates - the file should be extended with like [other curves](https://github.com/celer-network/icicle/blob/03fac235a420d73a40ce79e336e6ac0f6f21486c/icicle/utils/sharedmem.cuh#L260-L277)

vhnatyk replied:
like it's 4years+ but seems there is no better workaround for templated dynamic shared memory other than[ Lei Mao blog's](https://leimao.github.io/blog/CUDA-Shared-Memory-Templated-Kernel/)

ImmanuelSegol replied:
I think this worked, im building it now

DmytroTym replied:
But you haven't included `struct SharedMemory<BW6_671::scalar_t> {...}` and `struct SharedMemory<BW6_671::projective_t> {...}`, is it ok? @vhnatyk could you pls take a look again?

vhnatyk replied:
> while seems this may trick a linker - for compatibility at least with ntt templates - the file should be extended with like [other curves](https://github.com/celer-network/icicle/blob/03fac235a420d73a40ce79e336e6ac0f6f21486c/icicle/utils/sharedmem.cuh#L260-L277)

@ImmanuelSegol  - this may build with that stub function, but ntt should not pass correctly if shared mem kernel is used

vhnatyk replied:
pls remove as it just disables linking error and hides the issue
```
__device__ void Error_UnsupportedType() 
{
    // This function does not need a body, as it's just to throw a linker error.
}
```

and  add the actual implementation like

```
template <>
struct SharedMemory<BW6_671::scalar_t> {
  __device__ BW6_671::scalar_t* getPointer()
  {
    extern __shared__ BW6_671::scalar_t s_scalar_t_bw6_671[];
    return s_scalar_t_bw6_671;
  }
};

template <>
struct SharedMemory<BW6_671::projective_t> {
  __device__ BW6_671::projective_t* getPointer()
  {
    extern __shared__ BW6_671::projective_t s_projective_t_bw6_671[];
    return s_projective_t_bw6_671;
  }
};
#endif //_SHAREDMEM_H_
```

DmytroTym commented on goicicle/curves/bw6761/g2.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+type G2PointAffine struct {
+	X, Y ExtentionField
```
That's not how G2 works for this curve unfortunately, the coordinates of points are from the same field as G1 points and not extension field elements

ImmanuelSegol replied:
Done

DmytroTym commented on goicicle/curves/bw6761/g2.go:
```diff
@@ -0,0 +1,114 @@
+	return G2Point{
+		X: p.X,
+		Y: p.Y,
+		Z: ExtentionField{
```
In principle it's probably better to use functions like [projective_from_affine](https://github.com/ingonyama-zk/icicle/blob/main/icicle/curves/bls12_377/projective.cu#L34) from CUDA, it's better to have a single source of truth. Like in this case, because G2 is actually not over extension field, and the size of limbs is wrong too, this function might work incorrectly

ImmanuelSegol replied:
Thanks, this is a left over function that I didnt update.

ImmanuelSegol replied:
removed it

ImmanuelSegol committed:
Make dependency instructions more clear (#227)

jeremyfelder committed:
Fix: div by 0 when number of Elements is 1 (#230)

ImmanuelSegol committed:
update g2 size

ImmanuelSegol committed:
fix memory error in single_stage_multi_reduction_kernel (#235)

* refactor

* refactor

* revert

* refactor: clang format

* Update icicle/appUtils/msm/msm.cu

ImmanuelSegol committed:
refactor: fix curve

ImmanuelSegol committed:
refactor

ImmanuelSegol committed:
refactor

DmytroTym commented on icicle/curves/bw6_761/projective.cu:
```diff
@@ -64,10 +64,10 @@ extern "C" bool eq_bw6_761(BW6_761::projective_t* point1, BW6_761::projective_t*
+         !((point1->x == BW6_761::g2_point_field_t::zero()) && (point1->y == BW6_761::g2_point_field_t::zero()) &&
+           (point1->z == BW6_761::g2_point_field_t::zero())) &&
+         !((point2->x == BW6_761::g2_point_field_t::zero()) && (point2->y == BW6_761::g2_point_field_t::zero()) &&
+           (point2->z == BW6_761::g2_point_field_t::zero()));
```
Trying to `#include "projective.cu"` leads to a compile error as the compiler doesn't know about `g2_point_field_t` (you deleted it from `curve_config.cuh`). I think you should revert to using `point_field_t` here.

DmytroTym replied:
Well not `#include` but rather add to build

DmytroTym reviewed (changes requested):
Good job, I think you're almost done but I'd prefer that you delete newly auto-generated `src/curves/bw6_761.rs` and `src/test_bw6_761.rs`, they don't really work. Rust interface is not an immediate priority with this curve and we should introduce it as we refactor APIs on our feature branch IMO.

vhnatyk commented on goicicle/curves/bw6761/ntt_test.go:
```diff
@@ -0,0 +1,149 @@
+	NttBatch(&nttResult, false, count, 0)
+	assert.NotEqual(t, nttResult, scalars)
+
+	assert.Equal(t, nttResult, nttResult)
```
asserting the equality of the same objects?

vhnatyk commented on src/curves/mod.rs:
```diff
@@ -1,4 +1,3 @@
 pub mod bls12_377;
 pub mod bls12_381;
 pub mod bn254;
-pub mod bw6_761;
```
why is it removed?

DmytroTym replied:
I asked to remove it during the previous review as I removed the (very much incorrectly working) autogenerated Rust wrappers for BW6-671

vhnatyk reviewed (commented):
just a few issues in the comments should be addressed,  otherwise looks good 👍🏻

ImmanuelSegol committed:
refactor

DmytroTym committed:
Hotfix to go regression when 2 curves are imported (#245)

Hotfix to slowdown in go when more than one curve is imported.

ImmanuelSegol committed:
fix tests

jeremyfelder committed:
Exclude target directory from format checks (#247)

jeremyfelder committed:
[CI]: Add concurrency groups at workflow level (#238)

Add concurrency groups at workflow level for CI. Remove dev CI since we no longer use dev branch. Resolves #180

jeremyfelder committed:
Fix readme badge links, fix CI cpp formatter (#249)

DmytroTym committed:
Fixed issues preventing correctness for BW6-761

ImmanuelSegol committed:
Merge pull request #1 from DmytroTym/bw6/gowrapper

DmytroTym committed:
Merge remote-tracking branch 'ingo/main' into bw6/gowrapper

ImmanuelSegol committed:
Merge pull request #2 from DmytroTym/bw6/gowrapper

DmytroTym committed:
REAMDE updated

ImmanuelSegol committed:
Merge pull request #3 from DmytroTym/bw6/gowrapper
